### DISCUSSION TITLE
system wide installation, multiple python versions

### DISCUSSION THREAD

**Discussion by david-vicente** [python-lsp/python-lsp-server#304] [Q&A]
If I have installed `pylsp` with

``` shell
python -m pip install "python-lsp-server[all]"
```

using the system wide `python` interpreter (lets say version 3.10) that comes with my Linux distribution, but I want to use `pylsp` while editing a script for another python version (e.g. 3.11, installed using `pyenv`), is there a way to inform `pylsp` of this?

For example using `pyright` if I add 

``` toml
[tool.pyright]
pythonVersion = "3.11"
```
to `pyproject.toml` at the root of the project, `pyright` adapts accordingly, even though I also installed it using my system wide python running

``` shell
python -m pip install pyright
```

(I know that installing `pyright` through `pip` installs `node` in the background though...)

**Comment by doolio** [comment]
Are you saying that even after specifiying a `pyenv` installed python as the active version `pylsp` still relies on your system version?

FWIW, I use [`asdf`](https://github.com/asdf-vm/asdf) a language agnostic tool version manager to install python (it uses `pyenv` under the hood for python). With it I can set the active python at the system or local (read project) level. I then create a virtual environment based on the active python and install `pylsp` into the virtual environment.

**Reply by david-vicente** [reply]
When I opened the discussion my goal was to have `pylsp` installed globally and have it discover virtual environments and adapt to them. For example I wanted to install `pylsp` in my python3.10's global environment, and use that installation to edit projects that could be for example in python versions 3.9, 3.10 or 3.11. I didn't want to install `pylsp` in every virtual environment that I have.

After reading the documentations of `pylsp`'s dependencies I found out that:
- `jedi` allows me to achieve my goal, if I configure `pylsp.plugins.jedi.environment` from my client.
- `flake8` (which is a wrapper for `pycodestyle`, `maccabe` and `pyflakes`) does not allow me to achieve the goal because it has to be installed in the virtual environment of a python whose version is the same of the project that I'm editing. In other words, flake 8 does not care about virtual environments, but it does care about python versions.
- `pylint` does not allow me to achieve the goal because it has to be installed in the virtual environment of the project in order to use its full capabilities.
- `rope` I still haven't checked.

**Reply by david-vicente** [reply]
With `pyright` you can have a single `pyright` installed globally and if you use with `direnv` it will find the current active virtual environment and python version.

**Reply by david-vicente** [reply]
I can do the same with `pylsp` provided that I only enable `jedi`.

**Reply by doolio** [reply]
Interesting. Thanks for sharing your findings.

**Reply by david-vicente** [reply]
It seems that Rope also needs to be installed in each project's environment. See [this discussion](https://github.com/python-rope/rope/discussions/557).
